File: collect-strings.js
const collectStrings = (object) => {
  let stringsCollector = []

  function selectStrings (obj) {
    for(let key in obj) {
      if(typeof obj[key] === "string") {
        stringsCollector.push(obj[key])
      } else if(typeof obj[key] === "object") {
        selectStrings(obj[key])
      }
    }
  }

  selectStrings(object)

  return stringsCollector
}



const obj = {
  stuff: "foo",
  data: {
      val: {
          thing: {
              info: "bar",
              moreInfo: {
                  evenMoreInfo: {
                      weMadeIt: "baz"
                  }
              }
          }
      }
  }
}

console.log(collectStrings(obj)) // ["foo", "bar", "baz"])

File: collect_strings.rb
def collect_strings(object)
  strings_collector = []

  object.each do | key, value |
    if(value.kind_of?(String))
      strings_collector.push(value)
    end
  end

  strings_collector
end

obj = {
  stuff: "foo",
  data: {
    val: {
      thing: {
        info: "bar",
        moreInfo: {
          evenMoreInfo: {
            weMadeIt: "baz"
          }
        }
      }
    }
  }
}

p collect_strings